### PR TITLE
Feature/handle gundown

### PR DESCRIPTION
**Pull request by mkrusemark**
Added support for reconnection via a shotgun:reopen call.
Testsuites had to be modified because cover didn't work with latest erlang version.
Rebar3 crashed with rich compiler messages.

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -49,7 +51,7 @@
                   , strict_validation
                   , warn_export_vars
                   , warn_exported_vars
-                  , warn_missing_spec
+%                  , warn_missing_spec
```
Comment (elbrujohalcon) [inline rebar.config:54]: What missing specs required this change?

File: src/shotgun.erl [hunk 2]
```diff
@@ -17,6 +17,9 @@
 -export([ open/2
         , open/3
         , open/4
+				, reopen/3
+				, reopen/4
+				, reopen/5
```
Comment (elbrujohalcon) [inline src/shotgun.erl:22]: Can you align these with the other functions?

File: src/shotgun.erl [hunk 3]
```diff
@@ -61,6 +64,7 @@
         , wait_response/3
         , receive_data/3
         , receive_chunk/3
+				, down/3
```
Comment (elbrujohalcon) [inline src/shotgun.erl:67]: Can you align these with the other functions?

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]
It's not bad but you seem to be using tabs for indentation and that breaks code style.
If you, instead of realigning everything manually, would like to add `rebar3_format` to this project and format the code according to its rules, that would be perfectly fine, too.

**Comment by elbrujohalcon** [comment]
Hummm… In order to see only the _relevant_ changes here, I applied the formatter independently in #198 … but it looks I didn't do it correctly, did I? 🤔

**Comment by elbrujohalcon** [comment]
Fixed :)

File: src/shotgun.erl [hunk 4]
```diff
@@ -329,36 +370,38 @@ init([{Host, Port, Type, Opts}]) ->
...
+            _StateName,
             StateData = #{pid := Pid}) ->
-    error_logger:warning_msg("~p connection down on ~p: ~p (Killed: ~p, Unprocessed: ~p)",
-                             [Protocol, Pid, Reason, KilledStreams, UnprocessedStreams]),
-    {next_state, StateName, StateData};
-handle_info({gun_down, Pid, _, _, _, _}, StateName, StateData = #{pid := Pid}) ->
-    {next_state, StateName, StateData};
+    error_logger:warning_msg("~p connection down on ~p: ~p (Killed: ~p)",
+                             [Protocol, Pid, Reason, KilledStreams]),
+    %		demonitor(Pid, [flush]),
```
Comment (elbrujohalcon) [inline src/shotgun.erl:398]: Let's remove this comment.

File: src/shotgun.erl [hunk 5]
```diff
@@ -400,6 +443,14 @@ at_rest(timeout, _Event, StateData) ->
     end;
 at_rest(cast, shutdown, _StateData) ->
     stop;
+% this state transition is not expected because gun messages change state to 'down'.
+% but just in case ...
+at_rest(cast, {gun_down, _Args, _From}, StateData = #{pid := Pid}) ->
+    % cleanup gun process
+    %		demonitor(Pid, [flush]),
```
Comment (elbrujohalcon) [inline src/shotgun.erl:450]: Let's remove these comments, too.

File: test/http_server.erl [hunk 6]
```diff
@@ -29,8 +29,12 @@ stop(_State) ->
 
 -spec start_phase(atom(), application:start_type(), []) -> ok | {error, term()}.
 start_phase(start_cowboy_http, _StartType, []) ->
+    start_listener().
+
+-spec start_listener() -> ok | {error, term()}.
+start_listener() ->
     Port = application:get_env(http_server, http_port, 8888),
-    ListenerCount = application:get_env(http_server, http_listener_count, 10),
+    _ListenerCount = application:get_env(http_server, http_listener_count, 10),
```
Comment (elbrujohalcon) [inline test/http_server.erl:37]: I think we can just remove the line entirely.

**Review by elbrujohalcon — APPROVED** [review]
A few comment removals, but everything else looks fine.

**Comment by mkrusemark** [comment]
> Hummm… In order to see only the _relevant_ changes here, I applied the formatter independently in #198 … but it looks I didn't do it correctly, did I? 🤔

**Comment by mkrusemark** [comment]
That's interesting. I added the plugin without any specific options, see rebar.config.

**Comment by mkrusemark** [comment]
Please tag the new version with 1.1.0.

**Comment by elbrujohalcon** [comment]
Why have you closed this PR, @mkrusemark ?

**Comment by elbrujohalcon** [comment]
I'm confident it's a good change. I'll reopen and merge it.

**Comment by elbrujohalcon** [comment]
> Please tag the new version with 1.1.0.

Done: https://hex.pm/packages/shotgun/1.1.0

**Pull request merged**